editor/py/serve.py

```diff
@@ -662,85 +668,118 @@ def get_file(path):
-    """Returns a safe path in the editor filesystem.
+def get_user_path():
+    """Returns the path of the current user's temp directory.
 
     Uses flask.safe_join to check for directory traversal attacks.
 
-    Args:
-        file_name: Text filename.
-        user: The current user. If None, defaults to flask.g.user.
-        suffix: Text filename suffix. Defaults to '.pbtxt'.
-
     Returns:
-        Path to the requested file.
+        Path to the user directory.
     """
-    if not user:
-        user = flask.g.user
-    return flask.safe_join(app.config['ORD_EDITOR_LOCAL'], user,
-                           f'{file_name}{suffix}')
+    return flask.safe_join(TEMP, flask.g.user_id)
 
 
-def get_user_path():
-    """Checks that a username results in a valid path in the editor filesystem.
+def get_path(file_name, suffix=''):
+    """Returns a safe path in the user's temp directory.
 
     Uses flask.safe_join to check for directory traversal attacks.
 
+    Args:
+        file_name: Text filename.
+        suffix: Text filename suffix. Defaults to '.pbtxt'.
+
     Returns:
-        Path to the user directory.
+        Path to the requested file.
     """
-    return flask.safe_join(app.config['ORD_EDITOR_LOCAL'], flask.g.user)
+    return flask.safe_join(get_user_path(), f'{file_name}{suffix}')
+
+
+def exists_dataset(name):
+    """True if a dataset with the given name is defined for the current user."""
+    with flask.g.db.cursor() as cursor:
+        query = psycopg2.sql.SQL(
+            'SELECT 1 FROM datasets WHERE user_id=%s AND dataset_name=%s')
+        user_id = flask.g.user_id
+        cursor.execute(query, [user_id, name])
+        return cursor.rowcount > 0
+
+
+@app.route('/login')
+def show_login():
+    """Presents a form to set a new access token from a given user ID."""
+    return flask.render_template('login.html')
+
+
+@app.route('/authenticate', methods=['GET', 'POST'])
+def authenticate():
+    """Issue a new access token for a given user ID."""
+    # GET authentications always login as the test user.
+    if flask.request.method == 'GET':
+        return issue_access_token(TESTER)
+    user_id = flask.request.form.get('user_id')
+    if user_id is None or re.match('^[0-9a-fA-F]{32}$', user_id) is None:
+        return flask.redirect('/login')
+    return issue_access_token(user_id)
+
+
+def issue_access_token(user_id):
+    """Login as the given user and set the access token in a response."""
+    with flask.g.db.cursor() as cursor:
+        query = psycopg2.sql.SQL('INSERT INTO logins VALUES (%s, %s, %s)')
+        access_token = uuid.uuid4().hex
+        timestamp = int(time.time())
+        cursor.execute(query, [access_token, user_id, timestamp])
+        flask.g.db.commit()
+        response = flask.redirect('/')
+        # Expires in a year.
+        response.set_cookie('Access-Token', access_token, max_age=31536000)
+        return response
+
+
+def make_user(name='auto'):
```

Comment:
Please add a more complete docstring with Args

https://google.github.io/styleguide/pyguide.html#383-functions-and-methods

Reply:
Done.